### DISCUSSION TITLE
Issues implementing object dragging with Distance constraint

### DISCUSSION THREAD

**Discussion by tksuoran** [jrouwe/JoltPhysics#279] [Q&A]
I am trying to implement feature to drag objects with mouse / OpenXR controller.

I have a working version for Bullet, which uses Point constraint, and specific settings (constraint with tau = 0.001 and impulse clamp 1.0, dragged rigid body linear and angular damping overridden during dragging to value close to but slightly less than 1.0) to minimize oscillation.

The Jolt version currently uses Distance constraint together with a non-colliding rigid body for which I change world transform when the dragging point moves. This is functional, but I have been unable solve the following issues:

- Using frequency < 1.0, the dragging effect is too small / slow
- Using frequency >= 1.0, there is oscillation (both linear and angular), and attempts to override damping settings for the dragged body do not seem to have an effect.
- Using frequency = 0.0, the dragged object does not gain any kinetic energy when distance constraint is removed, and the dragged object is rotated more than I like, and I have not found a way to dampen this.

I would like to achieve smooth and fast but not quite instant dragging effect with minimal oscillation.
Is there some specific reason why I do not get damping for the dragged object?
Any advice is welcome - thanks!

**Comment by jrouwe** [comment]
In that case I think you should go for the other solution that I posted in #268, using a point constraint on a kinematic object, I've modified the sample to show you how that is done (aim at a body in the center of the screen and hold spacebar):

```
void SamplesApp::UpdateDebug()
{
	JPH_PROFILE_FUNCTION();

	const float cDragRayLength = 40.0f;

	BodyInterface &bi = mPhysicsSystem->GetBodyInterface();
			
	// Handle keyboard input for which simulation needs to be running
	for (int key = mKeyboard->GetFirstKey(); key != 0; key = mKeyboard->GetNextKey())
		switch (key)
		{
		case DIK_B:
			ShootObject();
			break;
		}

	// Allow the user to drag rigid bodies around
	if (mDragConstraint == nullptr)
	{
		// Not dragging yet
		Vec3 hit_position;
		float hit_fraction;
		if (CastProbe(cDragRayLength, hit_fraction, hit_position, mDragBody))
		{
			// If key is pressed create constraint to start dragging
			if (mKeyboard->IsKeyPressed(DIK_SPACE))
			{
				// Construct fixed body for the mouse constraint
				// Note that we don't add it to the world since we don't want anything to collide with it, we just
				// need an anchor for a constraint
				Body *drag_anchor = bi.CreateBody(BodyCreationSettings(new SphereShape(0.01f), hit_position, Quat::sIdentity(), EMotionType::Kinematic, Layers::MOVING));
				bi.AddBody(drag_anchor->GetID(), EActivation::Activate);
				mDragAnchor = drag_anchor;

				{
					// Target body must be dynamic
					BodyLockWrite lock(mPhysicsSystem->GetBodyLockInterface(), mDragBody);
					if (lock.Succeeded())
					{
						Body &drag_body = lock.GetBody();
						if (drag_body.IsDynamic())
						{
							// Create constraint to drag body
							PointConstraintSettings settings;
							settings.mPoint1 = settings.mPoint2 = hit_position;

							// Construct constraint that connects the drag anchor with the body that we want to drag
							mDragConstraint = settings.Create(*drag_anchor, drag_body);
							mPhysicsSystem->AddConstraint(mDragConstraint);

							mDragFraction = hit_fraction;
						}
					}
				}

				if (mDragConstraint == nullptr)
				{
					// No constraint created, remove body again
					bi.RemoveBody(mDragAnchor->GetID());
					bi.DestroyBody(mDragAnchor->GetID());
					mDragAnchor = nullptr;
				}
			}
		}
	}
	else
	{
		if (!mKeyboard->IsKeyPressed(DIK_SPACE))
		{
			// If key released, destroy constraint
			if (mDragConstraint != nullptr)
				mPhysicsSystem->RemoveConstraint(mDragConstraint);
			mDragConstraint = nullptr;

			// Destroy drag anchor
			bi.RemoveBody(mDragAnchor->GetID());
			bi.DestroyBody(mDragAnchor->GetID());
			mDragAnchor = nullptr;

			// Forget the drag body
			mDragBody = BodyID();
		}
		else
		{
			// Else update position of anchor
			bi.MoveKinematic(mDragAnchor->GetID(), GetCamera().mPos + cDragRayLength * mDragFraction * GetCamera().mForward, Quat::sIdentity(), 1.0f / mUpdateFrequency);

			// Activate other body
			bi.ActivateBody(mDragBody);
		}
	}
}
```

Note that in this implementation the small sphere that I use to attach the point constraint to collides with the sphere, which you should try to avoid by putting it in a non-colliding layer.
